static/app/views/dashboards/widgetBuilder/hooks/useQueryParamState.tsx

```diff
@@ -7,9 +7,22 @@ import useLocationQuery from 'sentry/utils/url/useLocationQuery';
 import {useLocation} from 'sentry/utils/useLocation';
 import {useNavigate} from 'sentry/utils/useNavigate';
 
-export function useQueryParamState(
-  fieldName: string
-): [string | undefined, (newField: string | undefined) => void] {
+interface UseQueryParamStateProps<T> {
+  fieldName: string;
+  decoder?: (value: string) => T;
+}
+
+/**
+ * Hook to manage a state that is synced with a query param in the URL
+ *
+ * @param fieldName - The name of the query param to sync with the state
+ * @param decoder - A function to decode the query param value into the desired type
+ * @returns A tuple containing the current state and a function to update the state
+ */
+export function useQueryParamState<T = string>({
+  fieldName,
+  decoder,
+}: UseQueryParamStateProps<T>): [T | undefined, (newField: T | undefined) => void] {
```

Comment:
I think in the future I'll have to label these as `serialize` and `deserialize`, because it needs to go from URL state into whatever types need to work with the components, and from those types back into URL state, but since display types are pretty much strings this is fine for now without the `serialize` side.